File: core/account/tx/handlers/swap/resolve/settlement.ts
import type { AccountState } from '../../../../../types/account';
import type { AccountDraftState } from '../../../../state/account-state-draft';
import { deriveDelta } from '../../../../utils';
import { createDefaultDelta } from '../../../../state/delta';
import { deriveSwapOffdeltaChanges } from '../../../../../orderbook/swap-execution';
import { getHold, releaseHold } from '../../../hold-utils';
import { createDeltaDraft } from '../../../delta-utils';
import { deriveTransferOffdeltaChange } from '../../../../../protocol/transform/delta-movement';
import { accountTxValidationRejected } from '../../../apply-result';
import type {
  AppliedSwapResolve,
  SwapResolveFailure,
  ValidatedSwapResolve,
} from './types';

const failure = (events: string[], error: string): SwapResolveFailure =>
  accountTxValidationRejected(error, events);

const validateCounterpartyCapacity = (
  account: AccountState,
  resolve: ValidatedSwapResolve,
  events: string[],
): SwapResolveFailure | null => {
  if (resolve.filledWant <= 0n) return null;
  const tokenId = resolve.offer.wantTokenId;
  const delta = account.deltas.get(tokenId) ?? createDefaultDelta(tokenId);
  const capacity = deriveDelta(delta, !resolve.offer.makerIsLeft).outCapacity;
  return resolve.filledWant > capacity
    ? failure(
        events,
        `Counterparty insufficient capacity on token ${tokenId}: ` +
        `needs ${resolve.filledWant}, has ${capacity}`,
      )
    : null;
};

export const applySwapResolveFinancials = (
  account: AccountDraftState,
  resolve: ValidatedSwapResolve,
  events: string[],
): AppliedSwapResolve | SwapResolveFailure => {
  const giveDelta = createDeltaDraft(account, resolve.offer.giveTokenId);
  const wantDelta = createDeltaDraft(account, resolve.offer.wantTokenId);
  const capacityFailure = validateCounterpartyCapacity(account, resolve, events);
  if (capacityFailure) return capacityFailure;

  const makerHoldSide = resolve.offer.makerIsLeft ? 'left' : 'right';
  const currentMakerHold = getHold(giveDelta, makerHoldSide);
  if (currentMakerHold < resolve.canonicalQuantizedGive) {
    return failure(
      events,
      `Hold underflow: current=${currentMakerHold} ` +
      `< required=${resolve.canonicalQuantizedGive}`,
    );
  }
  if (resolve.filledGive > 0n) {
    const change = deriveSwapOffdeltaChanges(
      resolve.offer.makerIsLeft,
      resolve.filledGive,
      resolve.filledWant,
    );
    giveDelta.offdelta += change.give;
    wantDelta.offdelta += change.want;
    events.push(
      `💱 Swap filled: ${resolve.filledGive} token${resolve.offer.giveTokenId} ` +
      `for ${resolve.filledWant} token${resolve.offer.wantTokenId}`,
    );
  }
  if (resolve.feeAmount > 0n) {
    wantDelta.offdelta += deriveTransferOffdeltaChange(
      resolve.offer.makerIsLeft,
      resolve.feeAmount,
    );
    events.push(`💸 Swap taker fee: ${resolve.feeAmount} token${resolve.effectiveFeeTokenId}`);
  }
  const releaseError = releaseHold(
    giveDelta,
    makerHoldSide,
    resolve.filledGive,
    (currentHold, releaseAmount) =>
      `Hold underflow: current=${currentHold} < required=${releaseAmount}`,
  );
  if (releaseError) return failure(events, releaseError);
  return { ...resolve, giveDelta, wantDelta, makerHoldSide };
};

File: rscore/crates/engine/src/swap/resolve/settlement.rs
//! Parity target: `core/account/tx/handlers/swap/resolve/settlement.ts`.
//!
//! Both token legs and the maker hold move together on the candidate: holds
//! and offdeltas are hashed Account state, so validation and commit run this
//! same mutation.

use num_bigint::BigInt;

use super::types::{AppliedSwapResolve, ValidatedSwapResolve};
use crate::error::ValidationRejection;
use crate::{AccountReplica, Delta, Side, TokenId, TransitionError};

fn rejected(code: &'static str) -> ValidationRejection {
    ValidationRejection::SwapResolve { code }
}

/// The counterparty pays the want leg, so its own out capacity bounds the fill.
fn validate_counterparty_capacity(
    want_delta: &Delta,
    resolve: &ValidatedSwapResolve,
) -> Result<(), ValidationRejection> {
    if resolve.filled_want <= BigInt::from(0) {
        return Ok(());
    }
    let taker = if resolve.offer.maker_is_left() {
        Side::Right
    } else {
        Side::Left
    };
    if resolve.filled_want > want_delta.perspective(taker).out_capacity {
        return Err(rejected("SWAP_RESOLVE_COUNTERPARTY_CAPACITY"));
    }
    Ok(())
}

pub(crate) struct SettledLegs {
    pub give_delta: Delta,
    pub want_delta: Delta,
    pub applied: AppliedSwapResolve,
    pub events: Vec<String>,
}

pub(crate) fn apply_swap_resolve_financials(
    replica: &AccountReplica,
    resolve: ValidatedSwapResolve,
) -> Result<Result<SettledLegs, ValidationRejection>, TransitionError> {
    let give_token = TokenId::new(resolve.offer.give_token_id())?;
    let want_token = TokenId::new(resolve.offer.want_token_id())?;
    let mut give_delta = replica.state().delta_or_zero(give_token)?;
    let mut want_delta = replica.state().delta_or_zero(want_token)?;
    if let Err(rejection) = validate_counterparty_capacity(&want_delta, &resolve) {
        return Ok(Err(rejection));
    }
    let maker_hold_side = if resolve.offer.maker_is_left() {
        Side::Left
    } else {
        Side::Right
    };
    if give_delta.hold(maker_hold_side) < &resolve.canonical_quantized_give {
        return Ok(Err(rejected("SWAP_RESOLVE_HOLD_UNDERFLOW")));
    }
    let mut events = Vec::new();
    let zero = BigInt::from(0);
    if resolve.filled_give > zero {
        // The maker sends give, the taker sends want.
        give_delta.apply_transfer(maker_hold_side, &resolve.filled_give)?;
        want_delta.apply_transfer(maker_hold_side.opposite(), &resolve.filled_want)?;
        events.push(format!(
            "💱 Swap filled: {} token{} for {} token{}",
            resolve.filled_give,
            resolve.offer.give_token_id(),
            resolve.filled_want,
            resolve.offer.want_token_id(),
        ));
    }
    if resolve.fee_amount > zero {
        // The fee moves on the want leg from the maker to the taker.
        want_delta.apply_transfer(maker_hold_side, &resolve.fee_amount)?;
        events.push(format!(
            "💸 Swap taker fee: {} token{}",
            resolve.fee_amount, resolve.effective_fee_token_id,
        ));
    }
    if give_delta.hold(maker_hold_side) < &resolve.filled_give {
        return Ok(Err(rejected("SWAP_RESOLVE_HOLD_UNDERFLOW")));
    }
    give_delta.release_hold(maker_hold_side, &resolve.filled_give)?;
    Ok(Ok(SettledLegs {
        give_delta,
        want_delta,
        applied: AppliedSwapResolve {
            resolve,
            maker_hold_side,
        },
        events,
    }))
}
